File: src/main/java/dev/qixils/quasicolon/decorators/option/AutoCompleteWith.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package dev.qixils.quasicolon.decorators.option;

import dev.qixils.quasicolon.decorators.functional.AutoCompleter;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Denotes the class that should be used to generate tab completions for a {@link Option}.
 * This is only to be used by options whose {@link Option#type() type}
 * {@link net.dv8tion.jda.api.interactions.commands.OptionType#canSupportChoices() supports choices}.
 */
@Target(ElementType.PARAMETER)
@Retention(RetentionPolicy.RUNTIME)
public @interface AutoCompleteWith {

	/**
	 * The class to use for generating auto-complete suggestions.
	 *
	 * @return auto-completer class
	 */
	Class<? extends AutoCompleter> value();
}

File: src/main/java/dev/qixils/quasicord/decorators/option/AutoCompleteWith.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package dev.qixils.quasicord.decorators.option

import dev.qixils.quasicord.autocomplete.AutoCompleter
import kotlin.reflect.KClass

/**
 * Denotes the class that should be used to generate tab completions for a [Option].
 * This is only to be used by options whose [type][Option.type]
 * [supports choices][net.dv8tion.jda.api.interactions.commands.OptionType.canSupportChoices].
 */
@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
annotation class AutoCompleteWith(

    /**
     * The class to use for generating auto-complete suggestions.
     *
     * @return auto-completer class
     */
    val value: KClass<out AutoCompleter>
)
